packages/babel-parser/src/parser/node.js

```diff
@@ -47,6 +48,46 @@ class Node implements NodeBase {
     return newNode;
   }
 }
+const NodePrototype = Node.prototype;
+
+function clonePlaceholder(node: any): any {
+  return cloneIdentifier(node);
+}
+
+export function cloneIdentifier(node: any): any {
```

Comment:
Unfortunately I can not get flow types working on this method. The `Placeholder` type complicates the typings here.

Reply:
Does using generics work? Something like
```js
export function cloneIdentifier<T: Identifier | Placeholder<"Identifier">>(node: T): T
```

Reply:
I tried generics. However it seems that Flow cannot infer the type from `type === "Placeholder"`. So I fallback to invariant comments, however Flow will throw "unexpected )" at the invariant comment `/*:: invariant(node instanceof Placeholder<"Identifier">) */`.

Reply:
Has the parser conversion to TS already started?

Reply:
Not yet; you can see the wip at https://github.com/babel/babel/pull/11578